### PR TITLE
Bump rust nightly version to 2021-05-27 to fix tock-registers

### PR DESCRIPTION
**Pull request by Cryptjar**
At least since the Rust nightly version (`070e459c2 2021-05-11`), the `tock-registers` crate fails to compile with the following message:

```text
Compiling tock-registers v0.6.0
error[E0557]: feature has been removed
 --> libraries/tock-register-interface/src/lib.rs:5:12
  |
5 | #![feature(const_fn)]
  |            ^^^^^^^^ feature has been removed
  |
  = note: split into finer-grained feature gates

error: aborting due to previous error
```

### Pull Request Overview

This PR fixes this by specializing the feature to `const_fn_trait_bound`.

### Testing Strategy

This pull request was tested by executing `cargo +nightly test` in `./libraries/tock-register-interface` folder, successfully.

### TODO or Help Wanted

There is an incompatibility barrier regarding the above feature depending on the toolchain selected. In the currently selected version (nightly-2021-03-19) by `./rust-toolchain`, the `#![feature(const_fn)]` does work, but `#![feature(const_fn_trait_bound)]` does not. On the other hand, if the toolchain would be updated, it would require some coordinated effort to update all the sub-crates in this repository, because as previously state the pervasively used `#![feature(const_fn)]` is not accepted by the latest toolchain (e.g. `070e459c2 2021-05-11`)

**Edit:** I did that toolchain update, it was mostly `s/consf_fn/const_fn_trait_bound/` (see 00be0e7b8bc4e6d9011e27c14438ad88597075b1)

### Documentation Updated

No changes.

### Formatting

- [x] Ran `make prepush`. ~~(**Fails**, see 'TODO or Help Wanted' above)~~ Works now.

### CODE REVIEW

**Comment by lschuermann** [comment]
Thanks. I think this is an important and mostly uncontroversial change. However, I'm not entirely sure whether there are additional processes we'd like to follow when bumping the Rust version. I remember there were some plans on when to do that specifically w.r.t. the upcoming Tock 2.0 release. @tock/core-wg any thoughts on this?

Also, there seems to be some other breakages in the `asm!` blocks in Tock. I'll happily look into those if you'd like. Would you mind me pushing some commits onto this, in particular to also update the `shell.nix` toolchain @Cryptjar?

**Comment by Cryptjar** [comment]
@lschuermann I also just noticed the `asm!` change, but I have no insights into the `tock` system, I'm just a down-stream `tock-registers` user. Please feel free to make any updates/changes.

**Comment by lschuermann** [comment]
It seems as of this version we can't clobber `r9` _in addition_ to `r6` and `r7` on Cortex-M. Hence I've used the same workaround, storing this register in a different one (`r12`) and clobbering that instead. I've yet to test this though. Given I'm more comfortable with RISC-V assembly, may someone more experienced verify these changes, e.g. @bradjc?

Also, I've taken this as an opportunity to make the Nix expression read the `rust-toolchain` file and use the channel and date specified there, so those two should stay in sync from now on.

**Comment by lschuermann** [comment]
Also (as a downstream user of `tock-registers` this might be of interest for you), we're going to release a new version soon, presumably containing these changes. You might want to keep an eye on #2579, I should have one of the PRs for that ready shortly.

**Comment by hudson-ayers** [comment]
Updating the rust version requires changing several other files, I added a commit to include those changes. I also updated `update_rust_version.sh` to no longer touch `shell.nix`, since Leon modified it to pull directly from `rust-toolchain`

**Comment by bradjc** [comment]
Did anyone run the update script? Do we know if 2021-05-12 is a valid nightly with all of the required tools?

**Comment by Cryptjar** [comment]
@bradjc I wasn't even aware of such a script. Also, I picked 2021-05-12 pretty much at random.

**Comment by bradjc** [comment]
```
./tools/update_rust_version.sh
Updating rustc to latest compatible version...
info: syncing channel updates for 'nightly-x86_64-apple-darwin'
info: latest update on 2021-05-21, rust version 1.54.0-nightly (40d230204 2021-05-20)
info: skipping nightly which is missing installed components 'miri', 'rls'
info: syncing channel updates for 'nightly-2021-05-20-x86_64-apple-darwin'
info: latest update on 2021-05-20, rust version 1.54.0-nightly (f94942d84 2021-05-19)
info: skipping nightly which is missing installed components 'miri', 'rls'
info: syncing channel updates for 'nightly-2021-05-19-x86_64-apple-darwin'
info: latest update on 2021-05-19, rust version 1.54.0-nightly (4e3e6db01 2021-05-18)
info: skipping nightly which is missing installed component 'rls'
info: syncing channel updates for 'nightly-2021-05-18-x86_64-apple-darwin'
info: latest update on 2021-05-18, rust version 1.54.0-nightly (3e99439f4 2021-05-17)
info: skipping nightly which is missing installed component 'miri'
info: syncing channel updates for 'nightly-2021-05-17-x86_64-apple-darwin'
info: latest update on 2021-05-17, rust version 1.54.0-nightly (fe72845f7 2021-05-16)
info: skipping nightly which is missing installed component 'rls'
info: syncing channel updates for 'nightly-2021-05-16-x86_64-apple-darwin'
info: latest update on 2021-05-16, rust version 1.54.0-nightly (8cf990c9b 2021-05-15)
info: skipping nightly which is missing installed component 'rls'
info: syncing channel updates for 'nightly-2021-05-15-x86_64-apple-darwin'
info: latest update on 2021-05-15, rust version 1.54.0-nightly (1025db84a 2021-05-14)
info: skipping nightly which is missing installed components 'rls', 'rustfmt'
info: syncing channel updates for 'nightly-2021-05-14-x86_64-apple-darwin'
info: latest update on 2021-05-14, rust version 1.54.0-nightly (6d395a1c2 2021-05-13)
info: skipping nightly which is missing installed components 'rls', 'rustfmt'
info: syncing channel updates for 'nightly-2021-05-13-x86_64-apple-darwin'
info: latest update on 2021-05-13, rust version 1.54.0-nightly (21e92b973 2021-05-12)
info: skipping nightly which is missing installed components 'rls', 'rustfmt'
info: syncing channel updates for 'nightly-2021-05-12-x86_64-apple-darwin'
info: latest update on 2021-05-12, rust version 1.54.0-nightly (5c0292654 2021-05-11)
info: skipping nightly which is missing installed component 'miri'
info: syncing channel updates for 'nightly-2021-05-11-x86_64-apple-darwin'
info: latest update on 2021-05-11, rust version 1.54.0-nightly (79e50bf77 2021-05-10)
info: downloading component 'cargo'
info: downloading component 'clippy'
info: downloading component 'llvm-tools-preview'
```

`2021-05-12` doesn't have miri. `2021-05-11` would work, however.

**Comment by Cryptjar** [comment]
`2021-05-11` would work for me :+1:

**Review by bradjc — DISMISSED** [review]

**Review by hudson-ayers — APPROVED** [review]

File: capsules/src/lib.rs [hunk 1]
```diff
@@ -1,4 +1,5 @@
-#![feature(const_fn)]
+#![feature(const_fn_trait_bound)]
+#![feature(const_fn_unsize)]
```
Comment (hudson-ayers) [inline capsules/src/lib.rs:2]: `const_fn_unsize` was stabilized two days ago, but I think its fine to stick with the slightly older nightly for now and remove this next time we update.

Comment (bradjc) [reply capsules/src/lib.rs:2]: Maybe we can give this one until Friday and see if Rust can get us a nightly with everything working before then?

Comment (hudson-ayers) [reply capsules/src/lib.rs:2]: That's fine too

**Comment by bradjc** [comment]
Hmm, not clear if a new nightly that works is going to happen any time soon: https://github.com/rust-lang/rust/issues/85453

**Comment by hudson-ayers** [comment]
> Hmm, not clear if a new nightly that works is going to happen any time soon: [rust-lang/rust#85453](https://github.com/rust-lang/rust/issues/85453)

Wait, why do we need rls?

**Comment by bradjc** [comment]
> > Hmm, not clear if a new nightly that works is going to happen any time soon: [rust-lang/rust#85453](https://github.com/rust-lang/rust/issues/85453)
> 
> Wait, why do we need rls?

Because the update rust version script says we do? Hard to say exactly who might be using it.

**Comment by ppannuto** [comment]
I believe it's because folks using IDEs (i.e. VS Code) rely on it and expect it to be there. It was an issue for some before that it was missing.

**Comment by ppannuto** [comment]
Let's stick with the plan above, give it till Friday to see if there's any movement on the RLS issue, then make the decision to merge or wait quickly on the call?

**Comment by Xanewok** [comment]
I'm hopeful that https://github.com/rust-lang/rust/pull/85721 lands before the upcoming nightly release, so that should fix the RLS breakage then.

**Review by bradjc — APPROVED** [review]
Ok great we have new nightly version that supports all of the tools, and let's us get rid of one of the new const_fn features!

**Review by lschuermann — APPROVED** [review]

**Review by hudson-ayers — APPROVED** [review]

**Pull request merged**